File: src/db/queries/investments.ts
import { getDatabase } from '../database';

export interface Investment {
  id: string;
  lender_id: string;
  trip_id: string;
  amount: number;
  interest_rate: number;
  expected_return: number;
  status: 'escrowed' | 'active' | 'completed' | 'defaulted';
  invested_at: string;
  maturity_date: string;
  completed_at?: string;
}

export interface CreateInvestmentInput {
  lender_id: string;
  trip_id: string;
  amount: number;
  interest_rate: number;
  expected_return: number;
  maturity_date: string;
}

/**
 * Get investment by ID
 */
export const getInvestment = (id: string): Investment | null => {
  const db = getDatabase();
  const stmt = db.prepare('SELECT * FROM investments WHERE id = ?');
  return stmt.get(id) as Investment | null;
};

/**
 * Get all investments
 */
export const getAllInvestments = (): Investment[] => {
  const db = getDatabase();
  const stmt = db.prepare('SELECT * FROM investments ORDER BY invested_at DESC');
  return stmt.all() as Investment[];
};

/**
 * Get investments by lender
 */
export const getInvestmentsByLender = (lenderId: string): Investment[] => {
  const db = getDatabase();
  const stmt = db.prepare('SELECT * FROM investments WHERE lender_id = ? ORDER BY invested_at DESC');
  return stmt.all(lenderId) as Investment[];
};

/**
 * Get investments by trip
 */
export const getInvestmentsByTrip = (tripId: string): Investment[] => {
  const db = getDatabase();
  const stmt = db.prepare('SELECT * FROM investments WHERE trip_id = ? ORDER BY invested_at DESC');
  return stmt.all(tripId) as Investment[];
};

/**
 * Get investments by status
 */
export const getInvestmentsByStatus = (status: Investment['status']): Investment[] => {
  const db = getDatabase();
  const stmt = db.prepare('SELECT * FROM investments WHERE status = ? ORDER BY invested_at DESC');
  return stmt.all(status) as Investment[];
};

/**
 * Get investments by lender and status
 */
export const getInvestmentsByLenderAndStatus = (lenderId: string, status: Investment['status']): Investment[] => {
  const db = getDatabase();
  const stmt = db.prepare('SELECT * FROM investments WHERE lender_id = ? AND status = ? ORDER BY invested_at DESC');
  return stmt.all(lenderId, status) as Investment[];
};

/**
 * Get investments by trip and lender
 */
export const getInvestmentsByTripAndLender = (tripId: string, lenderId: string): Investment[] => {
  const db = getDatabase();
  const stmt = db.prepare('SELECT * FROM investments WHERE trip_id = ? AND lender_id = ? ORDER BY invested_at DESC');
  return stmt.all(tripId, lenderId) as Investment[];
};

/**
 * Create investment
 */
export const createInvestment = (input: CreateInvestmentInput): Investment => {
  const db = getDatabase();
  const id = `inv-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;

  const stmt = db.prepare(`
    INSERT INTO investments (
      id, lender_id, trip_id, amount, interest_rate, expected_return, status, maturity_date
    ) VALUES (?, ?, ?, ?, ?, ?, 'escrowed', ?)
  `);

  stmt.run(
    id,
    input.lender_id,
    input.trip_id,
    input.amount,
    input.interest_rate,
    input.expected_return,
    input.maturity_date
  );

  const investment = getInvestment(id);
  if (!investment) {
    throw new Error('Failed to create investment');
  }

  return investment;
};

/**
 * Update investment status
 */
export const updateInvestmentStatus = (id: string, status: Investment['status']): Investment | null => {
  const db = getDatabase();

  const investment = getInvestment(id);
  if (!investment) return null;

  const updates = ['status = ?'];
  const values: any[] = [status];

  if (status === 'completed') {
    updates.push('completed_at = datetime(\'now\')');
  }

  values.push(id);

  const stmt = db.prepare(`
    UPDATE investments SET ${updates.join(', ')} WHERE id = ?
  `);

  stmt.run(...values);

  return getInvestment(id);
};

/**
 * Update investment
 */
export const updateInvestment = (id: string, updates: Partial<Investment>): Investment | null => {
  const db = getDatabase();

  const investment = getInvestment(id);
  if (!investment) return null;

  const fields: string[] = [];
  const values: any[] = [];

  Object.entries(updates).forEach(([key, value]) => {
    if (value !== undefined && key !== 'id') {
      fields.push(`${key} = ?`);
      values.push(value);
    }
  });

  if (fields.length === 0) return investment;

  values.push(id);

  const stmt = db.prepare(`
    UPDATE investments SET ${fields.join(', ')} WHERE id = ?
  `);

  stmt.run(...values);

  return getInvestment(id);
};

/**
 * Delete investment
 */
export const deleteInvestment = (id: string): boolean => {
  const db = getDatabase();
  const stmt = db.prepare('DELETE FROM investments WHERE id = ?');
  const result = stmt.run(id);
  return result.changes > 0;
};

/**
 * Get active investments count by lender
 */
export const getActiveInvestmentsCount = (lenderId: string): number => {
  const db = getDatabase();
  const stmt = db.prepare('SELECT COUNT(*) as count FROM investments WHERE lender_id = ? AND status = ?');
  const result = stmt.get(lenderId, 'active') as { count: number };
  return result.count;
};

/**
 * Get total invested amount by lender
 */
export const getTotalInvestedByLender = (lenderId: string): number => {
  const db = getDatabase();
  const stmt = db.prepare(`
    SELECT SUM(amount) as total FROM investments
    WHERE lender_id = ? AND status IN ('active', 'completed')
  `);
  const result = stmt.get(lenderId) as { total: number | null };
  return result.total || 0;
};

/**
 * Get total returns by lender
 */
export const getTotalReturnsByLender = (lenderId: string): number => {
  const db = getDatabase();
  const stmt = db.prepare(`
    SELECT SUM(expected_return) as total FROM investments
    WHERE lender_id = ? AND status = 'completed'
  `);
  const result = stmt.get(lenderId) as { total: number | null };
  return result.total || 0;
};

export default {
  getInvestment,
  getAllInvestments,
  getInvestmentsByLender,
  getInvestmentsByTrip,
  getInvestmentsByStatus,
  getInvestmentsByLenderAndStatus,
  getInvestmentsByTripAndLender,
  createInvestment,
  updateInvestmentStatus,
  updateInvestment,
  deleteInvestment,
  getActiveInvestmentsCount,
  getTotalInvestedByLender,
  getTotalReturnsByLender,
};

File: fastapi-backend/app/routers/investments.py
from fastapi import APIRouter, HTTPException, status, Depends
from typing import Optional
from psycopg2.extras import RealDictCursor
import time
import random
import string

from ..models import InvestmentResponse, InvestmentCreate, InvestmentStatusUpdate
from ..auth import get_current_user
from ..database import get_db_cursor

router = APIRouter(prefix="/investments", tags=["investments"])


def _convert_investment_to_float(inv):
    """Convert investment numeric fields from Decimal to float"""
    return {
        **inv,
        "amount": float(inv["amount"]),
        "interest_rate": float(inv["interest_rate"]),
        "expected_return": float(inv["expected_return"]),
    }


@router.get("/", response_model=list[InvestmentResponse])
async def get_investments():
    """Get all investments"""
    with get_db_cursor() as cursor:
        cursor.execute("SELECT * FROM investments ORDER BY invested_at DESC")
        investments = cursor.fetchall()

    return [_convert_investment_to_float(dict(inv)) for inv in investments]


@router.get("/{investment_id}", response_model=InvestmentResponse)
async def get_investment(investment_id: str):
    """Get investment by ID"""
    with get_db_cursor() as cursor:
        cursor.execute("SELECT * FROM investments WHERE id = %s", (investment_id,))
        investment = cursor.fetchone()

    if not investment:
        raise HTTPException(status_code=404, detail="Investment not found")

    return _convert_investment_to_float(dict(investment))


@router.get("/lender/{lender_id}", response_model=list[InvestmentResponse])
async def get_investments_by_lender(lender_id: str):
    """Get investments by lender"""
    with get_db_cursor() as cursor:
        cursor.execute("SELECT * FROM investments WHERE lender_id = %s ORDER BY invested_at DESC", (lender_id,))
        investments = cursor.fetchall()

    return [_convert_investment_to_float(dict(inv)) for inv in investments]


@router.get("/trip/{trip_id}", response_model=list[InvestmentResponse])
async def get_investments_by_trip(trip_id: str):
    """Get investments by trip"""
    with get_db_cursor() as cursor:
        cursor.execute("SELECT * FROM investments WHERE trip_id = %s ORDER BY invested_at DESC", (trip_id,))
        investments = cursor.fetchall()

    return [_convert_investment_to_float(dict(inv)) for inv in investments]


@router.post("/", response_model=InvestmentResponse, status_code=status.HTTP_201_CREATED)
async def create_investment(investment: InvestmentCreate):
    """Create new investment"""
    with get_db_cursor(commit=True) as cursor:
        inv_id = f"inv-{int(time.time() * 1000)}-{''.join(random.choices(string.ascii_lowercase + string.digits, k=9))}"

        cursor.execute(
            """
            INSERT INTO investments (
                id, lender_id, trip_id, amount, interest_rate, expected_return, status, maturity_date
            ) VALUES (%s, %s, %s, %s, %s, %s, 'escrowed', %s)
            RETURNING *
            """,
            (inv_id, investment.lender_id, investment.trip_id, investment.amount,
             investment.interest_rate, investment.expected_return, investment.maturity_date)
        )

        new_investment = cursor.fetchone()

    return _convert_investment_to_float(dict(new_investment))


@router.put("/{investment_id}/status", response_model=InvestmentResponse)
async def update_investment_status(investment_id: str, status_update: InvestmentStatusUpdate):
    """Update investment status"""
    with get_db_cursor(commit=True) as cursor:
        # Check if investment exists
        cursor.execute("SELECT id FROM investments WHERE id = %s", (investment_id,))
        if not cursor.fetchone():
            raise HTTPException(status_code=404, detail="Investment not found")

        if status_update.status == "completed":
            cursor.execute(
                "UPDATE investments SET status = %s, completed_at = NOW() WHERE id = %s RETURNING *",
                (status_update.status, investment_id)
            )
        else:
            cursor.execute(
                "UPDATE investments SET status = %s WHERE id = %s RETURNING *",
                (status_update.status, investment_id)
            )

        updated_investment = cursor.fetchone()

    return _convert_investment_to_float(dict(updated_investment))


@router.delete("/{investment_id}")
async def delete_investment(investment_id: str):
    """Delete investment"""
    with get_db_cursor(commit=True) as cursor:
        cursor.execute("DELETE FROM investments WHERE id = %s RETURNING id", (investment_id,))
        result = cursor.fetchone()

    if not result:
        raise HTTPException(status_code=404, detail="Investment not found")

    return {"message": "Investment deleted successfully"}
